samza-kafka/src/main/java/org/apache/samza/system/kafka/SamzaKafkaSystemAdmin.java

```diff
@@ -0,0 +1,680 @@
+import org.apache.samza.config.KafkaConfig;
+import org.apache.samza.config.SystemConfig;
+import org.apache.samza.system.ExtendedSystemAdmin;
+import org.apache.samza.system.StreamSpec;
+import org.apache.samza.system.StreamValidationException;
+import org.apache.samza.system.SystemStreamMetadata;
+import org.apache.samza.system.SystemStreamPartition;
+import org.apache.samza.util.ExponentialSleepStrategy;
+import org.apache.samza.util.ScalaJavaUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.Function0;
+import scala.Function1;
+import scala.Function2;
+import scala.Option;
+import scala.collection.JavaConverters;
+import scala.runtime.AbstractFunction0;
+import scala.runtime.AbstractFunction1;
+import scala.runtime.AbstractFunction2;
+import scala.runtime.BoxedUnit;
+
+import static org.apache.kafka.clients.consumer.KafkaConsumerConfig.*;
+
+
+public class SamzaKafkaSystemAdmin<K, V> implements ExtendedSystemAdmin {
+  private final String systemName;
+  private Consumer<K, V> metadataConsumer = null;
+
+  private final Supplier<ZkUtils> connectZk;
+  private final Supplier<AdminClient> connectAdminClient;
+
+  //Custom properties to use when the system admin tries to create a new
+  //coordinator stream.
+  private final Properties coordinatorStreamProperties;
+
+  //The replication factor to use when the system admin creates a new
+  // coordinator stream.
+  private final int coordinatorStreamReplicationFactor;
+
+  // Replication factor for the Changelog topic in kafka
+  //Kafka properties to be used during the Changelog topic creation
+  private final Map<String, ChangelogInfo> changelogTopicMetaInformation;
+
+  // Kafka properties to be used during the intermediate topic creation
+  private final Map<String, Properties> intermediateStreamProperties;
+
+  private static final Logger LOG = LoggerFactory.getLogger(SamzaKafkaSystemAdmin.class);
+
+  private final KafkaSystemAdminUtilsScala kafkaAdminUtils;
+
+  volatile private AdminClient adminClient = null;
+
+  private final boolean deleteCommittedMessages;
+
+  @VisibleForTesting
+  public static volatile boolean deleteMessageCalled = false;
+
+  // The same default exponential sleep strategy values as in open source
+  private static final double DEFAULT_EXPONENTIAL_SLEEP_BACK_OFF_MULTIPLIER = 2.0;
+  private static final long DEFAULT_EXPONENTIAL_SLEEP_INITIAL_DELAY_MS = 500;
+  private static final long DEFAULT_EXPONENTIAL_SLEEP_MAX_DELAY_MS = 10000;
+  private static final int MAX_RETRIES_ON_EXCEPTION = 5;
+
+  @Override
+  public void start() {
+    // Plese note. There is slight inconsistency in the use of this class.
+    // Some of the functinality of this class may actually be used BEFORE start() is called.
+    // The SamzaContainer gets metadata (using this class) in SamzaContainer.apply, but this "start" actually gets called in SamzaContainer.run.
+    // Also we assume that start is called only once.
+    if (metadataConsumer == null) {
```

Comment:
These if guards around `metadataConsumer`, `adminClient` are duplicated across in the ` start()`, `stop()` and individual API methods. Rather than duplicating these checks, we should add `Precondition.notNull` checks in the constructor once and delete it in the other individual API methods.